lib/openldap.c

```diff
@@ -1088,41 +1092,71 @@ static ssize_t oldap_recv(struct Curl_easy *data, int sockindex, char *buf,
 
           /* Binary value, encode to base64. */
           if(bvals[i].bv_len)
-            result = Curl_base64_encode(bvals[i].bv_val, bvals[i].bv_len,
-                                        &val_b64, &val_b64_sz);
-          if(!result)
-            result = client_write(data, STRCONST(": "), val_b64, val_b64_sz,
-                                  STRCONST("\n"));
+            *err = Curl_base64_encode(bvals[i].bv_val, bvals[i].bv_len,
+                                      &val_b64, &val_b64_sz);
+          if(!*err)
+            *err = client_write(data, STRCONST(": "), val_b64, val_b64_sz,
+                                STRCONST("\n"));
           free(val_b64);
         }
         else
-          result = client_write(data, STRCONST(" "),
-                                bvals[i].bv_val, bvals[i].bv_len,
-                                STRCONST("\n"));
-        if(result)
+          *err = client_write(data, STRCONST(" "),
+                              bvals[i].bv_val, bvals[i].bv_len,
+                              STRCONST("\n"));
+        if(*err)
           break;
       }
 
       ber_memfree(bvals);
       bvals = NULL;
-      if(!result)
-        result = client_write(data, STRCONST("\n"), NULL, 0, NULL, 0);
-      if(result)
+      if(!*err)
+        *err = client_write(data, STRCONST("\n"), NULL, 0, NULL, 0);
+      if(*err)
         break;
     }
 
     ber_free(ber, 0);
 
-    if(!result)
-      result = client_write(data, STRCONST("\n"), NULL, 0, NULL, 0);
-    if(!result)
-      result = CURLE_AGAIN;
+    if(!*err)
+      *err = client_write(data, STRCONST("\n"), NULL, 0, NULL, 0);
+    if(!*err)
+      *err = CURLE_AGAIN;
     break;
   }
 
   ldap_msgfree(msg);
-  *err = result;
-  return result? -1: 0;
+
+  return 1;
+}
+
+static ssize_t oldap_recv(struct Curl_easy *data, int sockindex, char *buf,
+                          size_t len, CURLcode *err)
+{
+  int read_count;
+  int msg_read;
+  (void)len;
+  (void)buf;
+  (void)sockindex;
+
+  read_count = 0;
+
+  /* There might be multiple messages so read all the available messages. */
+  do {
+    msg_read = oldap_recv_single(data, sockindex, buf, len, err);
+
+    if(msg_read)
+      read_count++;
+  } while(*err == CURLE_AGAIN && msg_read);
```

Comment:
Can you explain the logic here? Why does it loop because of this?

Reply:
Added to the description since it is related with the logic.